packages/compartment-mapper/test/conditional-host-exports.test.js

```diff
@@ -0,0 +1,40 @@
+import fs from 'fs';
+import url from 'url';
+import crypto from 'crypto';
+import 'ses';
+import test from 'ava';
+import { importLocation } from '../import.js';
+import { makeReadPowers } from '../node-powers.js';
+
+const fixture = new URL(
+  'fixtures-conditional-host-exports/node_modules/app/index.js',
+  import.meta.url,
+).toString();
+
+test('conditional host exports should fall through to ponyfill', async t => {
+  const readPowers = makeReadPowers({
+    url,
+    fs,
+    crypto,
+  });
+  const { namespace } = await importLocation(readPowers, fixture, {
+    conditions: new Set([]),
+  });
+  t.is(namespace.feature, 'ponyfill');
+});
+
+test('conditional host exports should exit to import hook', async t => {
+  const readPowers = makeReadPowers({
+    url,
+    fs,
+    crypto,
+  });
+  const { namespace } = await importLocation(readPowers, fixture, {
+    conditions: new Set(['endo:lib']),
```

Comment:
> This causes the module to be omitted from bundles ...

I'm trying to make such a bundle but losing:

```console
~/projects/endo/packages/compartment-mapper$ ../../node_modules/.bin/bundle-source -C endo:lib  ./test/fixtures-conditional-host-exports/node_modules/app/index.js 
(TypeError#1)
TypeError#1: Failed to load module "./index.js" in package "file:///home/connolly/projects/endo/packages/compartment-mapper/test/fixtures-conditional-host-exports/node_modules/app/" (1 underlying failures: Cannot find external module "endo:lib" in package file:///home/connolly/projects/endo/packages/compartment-mapper/test/fixtures-conditional-host-exports/node_modules/lib/

  at throwAggregateError (packages/ses/src/module-load.js:546:11)
  at load (packages/ses/src/module-load.js:594:3)
  at async digestFromMap (packages/compartment-mapper/src/archive-lite.js:364:3)
  at async makeAndHashArchiveFromMap (packages/compartment-mapper/src/archive-lite.js:415:52)
  at async bundleZipBase64 (packages/bundle-source/src/zip-base64.js:83:29)
  at async main (packages/bundle-source/src/main.js:107:20)
```

Reply:
I’ve got some more wrangling to do to make this work through the `makeBundle` to `parseBundle`, and a tiny amount of work to thread through from `bundleSource` to `importBundle` after that. Retreating to draft.

Reply:
This should work now. `bundle-source -C endo:lib -T` produced a bundle with a transparent exit import for me.

Reply:
confirmed:

```console
$ cd packages/compartment-mapper
$ mkdir -p exit
$ mkdir -p no-exit

$ ../../node_modules/.bin/bundle-source  -T ./test/fixtures-conditional-host-exports/node_modules/app/index.js | jq -r .endoZipBase64| base64 -d | bsdtar -C no-exit -xvf-
x compartment-map.json
x app/index.js
x lib/index.js

$ ../../node_modules/.bin/bundle-source -C endo:lib -T ./test/fixtures-conditional-host-exports/node_modules/app/index.js | jq -r .endoZipBase64| base64 -d | bsdtar -C exit -xvf-
x compartment-map.json
x app/index.js
x lib/endo.js

$ cat exit/lib/endo.js 
export { feature } from 'endo:lib';

$ cat no-exit/lib/index.js 
export const feature = 'ponyfill';
```

Reply:
## POC: Eliding @endo/patterns etc. from a contract bundle

ooh! works on `orca.contract.js` from [dapp-orchestration-basics](https://github.com/Agoric/dapp-orchestration-basics) too!

Contract bundle in its original form is around 2.1M:

```console
$ cd ~/projects/dapp-orchestration-basics/contract

$ ~/projects/endo/node_modules/.bin/bundle-source -T src/orca.contract.js | jq -r .endoZipBase64| base64 -d >no-exit.zip && ls -sh no-exit.zip 
2.1M no-exit.zip
```

Then I tweaked `@endo/patterns`, `@endo/far`, and `@endo/eventual-send` something like:

```
$ git diff .
diff --git a/packages/patterns/package.json b/packages/patterns/package.json
index b2b9866ea..1a149a52f 100644
--- a/packages/patterns/package.json
+++ b/packages/patterns/package.json
@@ -18,7 +18,10 @@
   "main": "./index.js",
   "module": "./index.js",
   "exports": {
-    ".": "./index.js",
+    ".": {
+      "endo:1": "./endo-eventual-send-exit.js",
+      "default": "./index.js"
+    },
     "./package.json": "./package.json"
   },
```

And tweaked the dapp dependencies:

```
$ ls -l ../node_modules/@endo/{far,patterns,eventual-send}
lrwxrwxrwx 1 connolly connolly 51 Aug 26 21:32 ../node_modules/@endo/eventual-send -> /home/connolly/projects/endo/packages/eventual-send
lrwxrwxrwx 1 connolly connolly 41 Aug 26 21:25 ../node_modules/@endo/far -> /home/connolly/projects/endo/packages/far
lrwxrwxrwx 1 connolly connolly 47 Aug 26 21:37 ../node_modules/@endo/patterns -> /home/connolly/projects/endo/packages/patterns/
```

Contract bundle shrinks to 17.M:

```
$ ~/projects/endo/node_modules/.bin/bundle-source -C endo:1 -T src/orca.contract.js | jq -r .endoZipBase64| base64 -d >exit.zip &&
 ls -sh exit.zip 
1.7M exit.zip
```